### PR TITLE
Add support for MoveCursor and UpdateCursor

### PR DESCRIPTION
**Pull request by dorindabassey**
Implement support for MoveCursor to set/show the cursor position and Implement support for UpdateCursor to update the cursor shape and location.

### CODE REVIEW

File: staging/vhost-device-gpu/src/virt_gpu.rs [hunk 1]
```diff
@@ -393,6 +394,28 @@ impl VirtioGpu {
...
     }
 
+    /// Moves the cursor's position to the given coordinates.
+    pub fn move_cursor(
+        &mut self,
+        resource_id: u32,
+        gpu_backend: &mut GpuBackend,
+        cursor: VhostUserGpuCursorPos,
+    ) -> VirtioGpuResult {
+        if resource_id == 0 {
```
Comment (mtjhrc) [inline staging/vhost-device-gpu/src/virt_gpu.rs:434]: This doesn't seem to follow the spec: 
> **VIRTIO_GPU_CMD_MOVE_CURSOR** Move cursor. Request data is struct virtio_gpu_update_cursor. Re-
sponse type is VIRTIO_GPU_RESP_OK_NODATA.
Move cursor to the place specified in pos. The other fields are not used and will be ignored by the
device.

Specifically the `The other fields are not used and will be ignored by the device.` is not followed.  You shouldn't check the `resource_id` here and call `cursor_pos_hide`. 


https://docs.oasis-open.org/virtio/virtio/v1.2/csd01/virtio-v1.2-csd01.pdf

File: staging/vhost-device-gpu/src/virt_gpu.rs [hunk 2]
```diff
@@ -393,6 +394,58 @@ impl VirtioGpu {
...
+            gpu_backend.cursor_pos_hide(&cursor).map_err(|e| {
+                error!("Failed to set cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+        } else {
+            gpu_backend.cursor_pos(&cursor).map_err(|e| {
+                error!("Failed to set cursor pos from frontend: {}", e);
+                ErrUnspec
+            })?;
+        }
```
Comment (mtjhrc) [inline staging/vhost-device-gpu/src/virt_gpu.rs:444]: This doesn't seem to follow the spec: 
> **VIRTIO_GPU_CMD_MOVE_CURSOR** Move cursor. Request data is struct virtio_gpu_update_cursor. Re-
sponse type is VIRTIO_GPU_RESP_OK_NODATA.
Move cursor to the place specified in pos. The other fields are not used and will be ignored by the
device.

Specifically the "The other fields are not used and will be ignored by the device." is not followed.  I don't think you should check the `resource_id` here and call `cursor_pos_hide`, this should probably only call `cursor_pos` .

File: staging/vhost-device-gpu/src/virt_gpu.rs [hunk 3]
```diff
@@ -393,6 +394,58 @@ impl VirtioGpu {
...
+
+        Ok(OkNoData)
+    }
```
Comment (mtjhrc) [inline staging/vhost-device-gpu/src/virt_gpu.rs:448]: This doesn't seem to follow the spec: 
> **VIRTIO_GPU_CMD_MOVE_CURSOR** Move cursor. Request data is struct virtio_gpu_update_cursor. Re-
sponse type is VIRTIO_GPU_RESP_OK_NODATA.
Move cursor to the place specified in pos. The other fields are not used and will be ignored by the
device.

Specifically the "The other fields are not used and will be ignored by the device." is not followed.  I don't think you should check the `resource_id` here and call `cursor_pos_hide`, this should probably only call `cursor_pos` .

File: staging/vhost-device-gpu/src/virt_gpu.rs [hunk 4]
```diff
@@ -393,6 +394,58 @@ impl VirtioGpu {
...
+        gpu_backend: &mut GpuBackend,
+        cursor_pos: VhostUserGpuCursorPos,
+        hot_x: u32,
+        hot_y: u32,
+        _mem: &GuestMemoryMmap,
+    ) -> VirtioGpuResult {
+        let size: usize = 4096;
+        let data = vec![0u32; size];
+        let mut data_array = [0u32; 4096];
+        data_array.copy_from_slice(&data[..]);
```
Comment (mtjhrc) [inline staging/vhost-device-gpu/src/virt_gpu.rs:412]: The `data_array.copy_from_slice` doesn't have any effect here -  you are copying data from a zero filled vector a zero filled array.  The `data` vector is then a practically unused variable. 

For a final implementation we need to copy the data  associated with `resource_id` into a temporary array witch is sent to the gpu_backend.  It is fine to leave the cursor image support not finished for now in this PR, just mentioning this.

File: staging/vhost-device-gpu/src/virt_gpu.rs (hunk 3 above)
Comment (dorindabassey) [reply staging/vhost-device-gpu/src/virt_gpu.rs:448]: this is a non-normative statement. the spec also says here that `Cursor will be loaded from the specified resource_id and will be moved to pos.` also they did not explicitly state what other fields will be ignored, becuase they could also be referring to the padding. Also in qemu depending if there is resource_id it does the following https://gitlab.com/qemu-project/qemu/-/blob/master/contrib/vhost-user-gpu/vhost-user-gpu.c#L969

File: staging/vhost-device-gpu/src/virt_gpu.rs (hunk 4 above)
Comment (dorindabassey) [reply staging/vhost-device-gpu/src/virt_gpu.rs:412]: yes, I agree I think I already added a todo below to not forget. I can add another todo for the `data` vec.

File: staging/vhost-device-gpu/src/virt_gpu.rs (hunk 3 above)
Comment (mtjhrc) [reply staging/vhost-device-gpu/src/virt_gpu.rs:448]: The sentence "Cursor will be loaded from the specified resource_id and will be moved to pos. " in spec is for `UPDATE_CURSOR` not `MOVE_CURSOR `.  

The spec is quite unclear, but if QEMU does this, then it should be fine

**Comment by mtjhrc** [comment]
Looks good, you can merge this. Just maybe add the TODO comment for CursorUpdate like we talked about.

**Pull request merged**